File: leetcode/number-of-islands/solution_test.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", ".."))
from typing import List
from tools.leetcode import test, summary
from solution import Solution
from solution2 import Solution as SolutionUF


def rows(*lines: str) -> List[List[str]]:
    """Fresh grid per call — numIslands mutates its input."""
    return [list(line) for line in lines]


def snake(n: int) -> List[List[str]]:
    """One island shaped as a single ~n*n/2 long corridor.

    Recursive DFS bottoms out at Python's 1000-frame limit on this input.
    """
    grid = [["0"] * n for _ in range(n)]
    for i in range(n):
        if i % 2 == 0:
            for j in range(n):
                grid[i][j] = "1"
        else:
            grid[i][n - 1 if (i // 2) % 2 == 0 else 0] = "1"
    return grid


for name, sol in (("dfs", Solution()), ("union-find", SolutionUF())):
    test(sol.numIslands(rows("11110", "11010", "11000", "00000")), 1, f"[{name}] Example 1")
    test(sol.numIslands(rows("11000", "11000", "00100", "00011")), 3, f"[{name}] Example 2")
    test(sol.numIslands(rows("1")), 1, f"[{name}] Single land cell")
    test(sol.numIslands(rows("0")), 0, f"[{name}] Single water cell")
    test(sol.numIslands(rows("101", "010", "101")), 5, f"[{name}] Diagonals are not connected")
    test(sol.numIslands(rows("1111111111")), 1, f"[{name}] Single row")
    test(sol.numIslands([["1"], ["0"], ["1"]]), 2, f"[{name}] Single column")
    test(sol.numIslands(snake(300)), 1, f"[{name}] 300x300 snake (no recursion limit)")

summary()

File: leetcode/number-of-islands/solution_test.cpp
#include "../../tools/leetcode.h"
#include "solution.cpp"

// Fresh grid per call — numIslands mutates its input.
vector<vector<char>> rows(vector<string> lines) {
    vector<vector<char>> grid;
    for (const string& line : lines) grid.push_back(vector<char>(line.begin(), line.end()));
    return grid;
}

// One island shaped as a single ~n*n/2 long corridor: the input that makes a
// recursive flood fill recurse ~45,000 deep at n = 300.
vector<vector<char>> snake(int n) {
    vector<vector<char>> grid(n, vector<char>(n, '0'));
    for (int i = 0; i < n; i++) {
        if (i % 2 == 0) grid[i].assign(n, '1');
        else grid[i][(i / 2) % 2 == 0 ? n - 1 : 0] = '1';
    }
    return grid;
}

int main() {
    Solution sol;

    auto g1 = rows({"11110", "11010", "11000", "00000"});
    test(sol.numIslands(g1) == 1, "Example 1");

    auto g2 = rows({"11000", "11000", "00100", "00011"});
    test(sol.numIslands(g2) == 3, "Example 2");

    auto g3 = rows({"1"});
    test(sol.numIslands(g3) == 1, "Single land cell");

    auto g4 = rows({"0"});
    test(sol.numIslands(g4) == 0, "Single water cell");

    auto g5 = rows({"101", "010", "101"});
    test(sol.numIslands(g5) == 5, "Diagonals are not connected");

    auto g6 = rows({"1111111111"});
    test(sol.numIslands(g6) == 1, "Single row");

    auto g7 = rows({"1", "0", "1"});
    test(sol.numIslands(g7) == 2, "Single column");

    auto g8 = snake(300);
    test(sol.numIslands(g8) == 1, "300x300 snake (deep traversal)");

    return summary();
}
